src/OpenTelemetry.Instrumentation.AspNet/AspNetInstrumentationOptions.cs

```diff
@@ -35,8 +35,13 @@ public class AspNetInstrumentationOptions
         });
 
         /// <summary>
-        /// Gets or sets a hook to exclude calls based on domain or other per-request criterion.
+        /// Gets or sets a Filter function to filter instrumentation for requests on a per request basis.
+        /// The Filter gets the HttpContext, and should return a boolean.
+        /// If Filter returns true, the request is collected.
+        /// If Filter returns false, the request is filtered out.
+        /// If Filter throws exception, then it is considered as if no filter is configured
+        /// and request is collected.
         /// </summary>
-        internal Predicate<HttpContext> RequestFilter { get; set; }
+        public Func<HttpContext, bool> Filter { get; set; }
```

Comment:
Repasting the comment for discussion:

User configured a filter.
Filter throws ex.
We cannot let the original request fails, so we have cover up the user's mistake. So we catch it and log it.

The next question is: What about the fallback behavior when users filter threw? We cannot know their original intention anyway. We can either

1. Move on as if filter functions was not configured at all. (This is the current PR)
2. Assume that user wanted to filter out and don't collect the request.

@alanwest thinks we should use 2. As people may have configured "filter" to filter out some requests due to securuity concerns. 

I am also leaning towards option 2. User configured filter because user want to filter out "something". So if the filter crashed, we should filter out, even though we cannot know the exact request which user wanted out.

@reyang @CodeBlanch to share their ideas.

Reply:
Tough one! We should default to safe behavior. So if we're thinking people will want to filter out things for security, let's go with option 2?

Reply:
Ok.
So the description gets simpler:
If the Filter returns false or throws exception, the request is filtered out.

Reply:
Yeah, this aligns with the common sense - only if filter()=>true, item will be picked.

Reply:
I will merge this now. and open an item for discussion - Mikel thinks an Enum instead of bool may make better clarity here. Lets chat in a new issue I will open soon.

Reply:
Yep, Enum seems to be more aligned with C# style. filter=>bool is more lisp/scheme/node style.

Reply:
What I said to @cijothomas on Gitter was (basically) something about `if (Filter(request) == false) filterRequest() else collectRequest()` seems off?

Reply:
I have different understanding probably due to my background:
1. `filter` came from the lambda calculus, and has a root in lisp/scheme/node.js
    * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter
    * https://stackoverflow.com/questions/9458982/how-is-filter-implemented
2. In SQL or Linq, the idiomatic way is to use `Where <condition>` 
3. In English (natural language), I guess filter means "filter out" by default?

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/issues/1186 Lets continue this discussion in the new issue I opened. 
Merged this PR to make progress.

Reply:
Yep, no need to block on this - naming things is the hardest coding problem 😄

Reply:
@reyang Your logic makes sense. When I made the filter public in http instrumentation I originally changed it as it is here. But then I put it back. For reasons I can't put into words 😄 It just seemed less confusing. I could be in the minority on this.

Reply:
Reminded me of `memcpy`, which one is intuitive (and my answer is always "neither"):
* memcpy(source, destination, size)
* memcpy(destination, source, size)